### PR TITLE
Reduce code duplication in Forum Notification Mails

### PR DESCRIPTION
**Pull request by legionth**
Removes the code duplication in the notification mails of the forum.

### CODE REVIEW

**Comment by mjansenDatabay** [comment]
I suggest to add typehints and return type declarations to all new methods.

**Review by mjansenDatabay — CHANGES_REQUESTED** [review]
I suggest to add typehints and return type declarations to all new methods.

File: Modules/Forum/classes/class.ilForumMailNotification.php [hunk 1]
```diff
@@ -644,4 +295,138 @@ private function getSecurePostMessage()
...
+	 * Add body and send mail without attachments
+	 *
+	 * @param string $subject - Language id of subject
+	 * @param int $userId - id of the user recipient of the mail
+	 * @param string $customText - mail text after salutation
+	 * @param string $action - Language id of action
+	 * @param string|null $date - date to be added in mail
+	 */
+	private function addBodyToMailWithoutAttachments(
+		string $subject,
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumMailNotification.php]: @legionth I suggest to rename this to `$subjectLanguageVariable` oder `$subjectLanguageId`

Furthermore you pass the **subject** here, but the method is named `addBodyToMailWithoutAttachments`

File: Modules/Forum/classes/class.ilForumMailNotification.php [hunk 2]
```diff
@@ -644,4 +295,138 @@ private function getSecurePostMessage()
...
+		$this->addMailBody($subject, $userId, $customText, $action, $date);
+		$this->appendAttachments();
+		$this->addLinkToMail();
+		$this->sendMail(array($userId), array('system'));
+	}
+
+	/**
+	 * Add body and send mail without attachments
+	 *
+	 * @param string $subject - Language id of subject
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumMailNotification.php]: @legionth I suggest to rename this to `$subjectLanguageVariable` oder `$subjectLanguageId`

File: Modules/Forum/classes/class.ilForumMailNotification.php [hunk 3]
```diff
@@ -644,4 +295,138 @@ private function getSecurePostMessage()
...
+	private function addMailBody(
+		string $subject,
+		int $userId,
+		string $customText,
+		string $action,
+		string $date
+	) {
+		$date = $this->createMailDate($date);
+
+		$this->addMailSubject($subject);
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumMailNotification.php:362]: @legionth You set the **subject** here, but the method is named `addMailBody`

File: Modules/Forum/classes/class.ilForumMailNotification.php [hunk 4]
```diff
@@ -644,4 +295,138 @@ private function getSecurePostMessage()
...
+	 * Add body and send mail with attachments
+	 *
+	 * @param string $subject - Language id of subject
+	 * @param int $userId - id of the user recipient of the mail
+	 * @param string $customText - mail text after salutation
+	 * @param string $action - Language id of action
+	 * @param string|null $date - date to be added in mail
+	 */
+	private function addBodyToMailWithAttachments(
+		string $subject,
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumMailNotification.php]: @legionth I suggest to rename this to `$subjectLanguageVariable` oder `$subjectLanguageId`

Furthermore you pass the **subject** here, but the method is named `addBodyToMailWithoutAttachments`

File: Modules/Forum/classes/class.ilForumMailNotification.php [hunk 5]
```diff
@@ -644,4 +295,138 @@ private function getSecurePostMessage()
 		}
 		return strip_tags($pos_message);
 	}
-}
\ No newline at end of file
+
+	/**
+	 * Add body and send mail with attachments
+	 *
+	 * @param string $subject - Language id of subject
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumMailNotification.php]: @legionth I suggest to rename this to `$subjectLanguageVariable` oder `$subjectLanguageId`

File: Modules/Forum/classes/class.ilForumMailNotification.php [hunk 6]
```diff
@@ -128,421 +128,72 @@ public function send()
...
-						$this->appendBody(strip_tags($pos_message) . "\n");
-					}
-					$this->appendBody("------------------------------------------------------------\n");
-
-					$this->appendBody($this->getPermanentLink(self::PERMANENT_LINK_FORUM));
-					$this->appendBody(ilMail::_getInstallationSignature());
-
-					$this->sendMail(array($rcp), array('system'));
+					$customText = sprintf($this->getLanguageText('thread_deleted_by'), $ilUser->getLogin(),  $this->provider->getForumTitle());
+					$this->addBodyToMailWithoutAttachments('frm_noti_subject_del_thread', $rcp, $customText, 'content_deleted_thread');
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumMailNotification.php]: @legionth I recommend to case the arguments ((int) / (string) etc.)

File: Modules/Forum/classes/class.ilForumMailNotification.php [hunk 7]
```diff
@@ -128,421 +128,72 @@ public function send()
...
-
-					$this->appendAttachments();
-
-					$this->appendBody($this->getPermanentLink());
-					$this->appendBody(ilMail::_getInstallationSignature());
-
-					$this->sendMail(array($rcp), array('system'));
-
+					$customText = sprintf($this->getLanguageText('frm_noti_new_post'), $this->provider->getForumTitle());
+					$this->addBodyToMailWithAttachments('frm_noti_subject_new_post', $rcp, $customText, 'new_post');
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumMailNotification.php]: @legionth I recommend to case the arguments ((int) / (string) etc.)

File: Modules/Forum/classes/class.ilForumMailNotification.php [hunk 8]
```diff
@@ -128,421 +128,72 @@ public function send()
...
-					$this->appendBody("------------------------------------------------------------\n");
-
-					$this->appendAttachments();
-
-					$this->appendBody($this->getPermanentLink());
-					$this->appendBody(ilMail::_getInstallationSignature());
-
-					$this->sendMail(array($rcp), array('system'));
+					$customText = $this->getLanguageText('forums_post_activation_mail');
+					$this->addBodyToMailWithAttachments('frm_noti_subject_act_post', $rcp, $customText, 'new_post');
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumMailNotification.php]: @legionth I recommend to case the arguments ((int) / (string) etc.)

File: Modules/Forum/classes/class.ilForumMailNotification.php [hunk 9]
```diff
@@ -128,421 +128,72 @@ public function send()
...
-
-					$this->appendAttachments();
-
-					$this->appendBody($this->getPermanentLink());
-					$this->appendBody(ilMail::_getInstallationSignature());
-
-					$this->sendMail(array($rcp), array('system'));
-
+					$customText = $this->getLanguageText('forum_post_replied');
+					$this->addBodyToMailWithAttachments('frm_noti_subject_answ_post', $rcp, $customText, 'new_post');
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumMailNotification.php]: @legionth I recommend to case the arguments ((int) / (string) etc.)

File: Modules/Forum/classes/class.ilForumMailNotification.php [hunk 10]
```diff
@@ -128,421 +128,72 @@ public function send()
...
-					$this->appendBody("------------------------------------------------------------\n");
-
-					$this->appendAttachments();
-
-					$this->appendBody($this->getPermanentLink());
-					$this->appendBody(ilMail::_getInstallationSignature());
-
-					$this->sendMail(array($rcp), array('system'));
+					$customText = sprintf($this->getLanguageText('post_updated_by'), $this->provider->getPostUpdateUserName($this->getLanguage()), $this->provider->getForumTitle());
+					$this->addBodyToMailWithAttachments('frm_noti_subject_upt_post', $rcp, $customText, 'content_post_updated');
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumMailNotification.php]: @legionth I recommend to case the arguments ((int) / (string) etc.)

File: Modules/Forum/classes/class.ilForumMailNotification.php [hunk 11]
```diff
@@ -128,421 +128,72 @@ public function send()
...
-					$this->appendBody("------------------------------------------------------------\n");
-
-					$this->appendAttachments();
-
-					$this->appendBody($this->getPermanentLink());
-					$this->appendBody(ilMail::_getInstallationSignature());
-
-					$this->sendMail(array($rcp), array('system'));
+					$customText = sprintf($this->getLanguageText('post_censored_by'), $this->provider->getPostUpdateUserName($this->getLanguage()) ,$this->provider->getForumTitle());
+					$this->addBodyToMailWithAttachments('frm_noti_subject_cens_post', $rcp, $customText, 'content_censored_post');
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumMailNotification.php]: @legionth I recommend to case the arguments ((int) / (string) etc.)

File: Modules/Forum/classes/class.ilForumMailNotification.php [hunk 12]
```diff
@@ -128,421 +128,72 @@ public function send()
...
-
-					$this->appendAttachments();
-
-					$this->appendBody($this->getPermanentLink());
-					$this->appendBody(ilMail::_getInstallationSignature());
-
-					$this->sendMail(array($rcp), array('system'));
+					$customText = sprintf($this->getLanguageText('post_uncensored_by'), $this->provider->getPostUpdateUserName($this->getLanguage()));
+					$date = $this->provider->getPostCensoredDate();
+					$this->addBodyToMailWithAttachments('frm_noti_subject_uncens_post', $rcp, $customText, 'forums_the_post', $date);
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumMailNotification.php]: @legionth I recommend to case the arguments ((int) / (string) etc.)

File: Modules/Forum/classes/class.ilForumMailNotification.php [hunk 13]
```diff
@@ -128,421 +128,72 @@ public function send()
...
-						$this->appendBody(strip_tags($pos_message) . "\n");
-					}
-					$this->appendBody("------------------------------------------------------------\n");
-
-					$this->appendBody($this->getPermanentLink(self::PERMANENT_LINK_FORUM));
-					$this->appendBody(ilMail::_getInstallationSignature());
-
-					$this->sendMail(array($rcp), array('system'));
+					$customText = sprintf($this->getLanguageText('post_deleted_by'), $ilUser->getLogin(),  $this->provider->getForumTitle());
+					$this->addBodyToMailWithoutAttachments('frm_noti_subject_del_post', $rcp, $customText, 'content_deleted_post');
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumMailNotification.php]: @legionth I recommend to case the arguments ((int) / (string) etc.)

File: Modules/Forum/classes/class.ilForumMailNotification.php [hunk 14]
```diff
@@ -644,4 +295,138 @@ private function getSecurePostMessage()
...
+	 * @internal
+	 *
+	 * @param string $date
+	 * @return string
+	 */
+	private function createMailDate(string $date) : string
+	{
+		ilDatePresentation::setLanguage($this->language);
+
+		if ($date === null) {
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumMailNotification.php]: @legionth Is it a valid case that `$date` equals `null`. Do we have to change the primitive type hint in the method signature?

File: Modules/Forum/classes/class.ilForumMailNotification.php (hunk 3 above)
Comment (legionth) [reply Modules/Forum/classes/class.ilForumMailNotification.php:362]: Renamed the method

**Comment by legionth** [comment]
Renamed the methods in the latest commit and fixed your remarks @mjansenDatabay

File: Modules/Forum/classes/class.ilForumMailNotification.php [hunk 15]
```diff
@@ -128,421 +128,72 @@ public function send()
...
-					$this->appendBody("------------------------------------------------------------\n");
-
-					$this->appendAttachments();
-
-					$this->appendBody($this->getPermanentLink());
-					$this->appendBody(ilMail::_getInstallationSignature());
-
-					$this->sendMail(array($rcp), array('system'));
+					$customText = sprintf($this->getLanguageText('post_censored_by'), $this->provider->getPostUpdateUserName($this->getLanguage()) ,$this->provider->getForumTitle());
+					$this->sendMailWithAttachments('frm_noti_subject_cens_post', (int) $rcp, (string) $customText, 'content_censored_post');
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumMailNotification.php]: @legionth Maybe the date is missing here

File: Modules/Forum/classes/class.ilForumMailNotification.php [hunk 16]
```diff
@@ -128,421 +128,72 @@ public function send()
...
-					$this->appendBody("------------------------------------------------------------\n");
-
-					$this->appendAttachments();
-
-					$this->appendBody($this->getPermanentLink());
-					$this->appendBody(ilMail::_getInstallationSignature());
-
-					$this->sendMail(array($rcp), array('system'));
+					$customText = sprintf($this->getLanguageText('post_updated_by'), $this->provider->getPostUpdateUserName($this->getLanguage()), $this->provider->getForumTitle());
+					$this->sendMailWithAttachments('frm_noti_subject_upt_post', (int) $rcp, (string) $customText, 'content_post_updated');
```
Comment (mjansenDatabay) [inline Modules/Forum/classes/class.ilForumMailNotification.php]: @legionth The date is missing here

**Review by mjansenDatabay — APPROVED** [review]

**Pull request merged**